Add support for repeated XML elements without a name attribute

amoerie opened the pull request:
Fixes #36541
Fixes #36561 

This pull request adds support in Microsoft.Extensions.Configuration.Xml for repeated XML elements without requiring a Name attribute.
This solves a particularly subtle bug when configuring Serilog from an XML configuration source. For a full description, see https://github.com/dotnet/runtime/issues/36541

The original implementation of the XmlStreamConfigurationProvider has been modified to do the following:

- Maintain a stack of encountered XML elements while traversing the XML source. This is needed to detect siblings.
- When siblings are detected, automatically append an index to the generated configuration keys. This makes it work exactly the same as the JSON configuration provider with JSON arrays.

Tests are updated to reflect the new behavior:
- Tests that verified an exception occurs when entering duplicate keys have been removed. Duplicate keys are supported now.
- Add tests that verify duplicate keys result in the correct configuration keys, with all the lower/upper case variants and with support for the special "Name" attribute handling.

Thank you for your consideration.

amoerie committed:
Add support for repeated XML elements without a name attribute

This commit adds support in Microsoft.Extensions.Configuration.Xml for repeated XML elements without requiring a Name attribute.
This solves a particularly subtle bug when configuring Serilog from an XML configuration source. For a full description, see https://github.com/dotnet/runtime/issues/36541

The original implementation of the XmlStreamConfigurationProvider has been modified to do the following:

- Maintain a stack of encountered XML elements while traversing the XML source. This is needed to detect siblings.
- When siblings are detected, automatically append an index to the generated configuration keys. This makes it work exactly the same as the JSON configuration provider with JSON arrays.

Tests are updated to reflect the new behavior:
- Tests that verified an exception occurs when entering duplicate keys have been removed. Duplicate keys are supported now.
- Add tests that verify duplicate keys result in the correct configuration keys, with all the lower/upper case variants and with support for the special "Name" attribute handling.

amoerie committed:
Add missing null check in XmlConfigurationElement.cs

amoerie committed:
Disable duplicate key tests only for XML sources

amoerie committed:
Apply code style in ConfigurationProviderXmlTest

amoerie commented:
I could use a little help with the failing build, I don't see how it's related to my changes.

maryamariyan commented:
The failing build 

```
  Discovering: Microsoft.Extensions.Hosting.Unit.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  Microsoft.Extensions.Hosting.Unit.Tests (found 84 test cases)
  Starting:    Microsoft.Extensions.Hosting.Unit.Tests (parallel test collections = on, max threads = 2)
    Microsoft.Extensions.Hosting.Internal.HostTests.BackgroundServiceAsyncExceptionGetsLogged [FAIL]
      'BackgroundService failed' did not get logged
      Expected: True
      Actual:   False
      Stack Trace:
        /_/src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs(1256,0): at Microsoft.Extensions.Hosting.Internal.HostTests.BackgroundServiceAsyncExceptionGetsLogged()
[40m[32minfo[39m[22m[49m: Microsoft.Extensions.Hosting.Tests.HostTests[0]
      Request starting
  Finished:    Microsoft.Extensions.Hosting.Unit.Tests
=== TEST EXECUTION SUMMARY ===
   Microsoft.Extensions.Hosting.Unit.Tests  Total: 84, Errors: 0, Failed: 1, Skipped: 0, Time: 15.317s
```

is tracked by https://github.com/dotnet/runtime/issues/43389

maryamariyan commented:
Thanks @amoerie for porting over this change to dotnet/runtime. 

We'll need to do a full review on the changes as it seems like there is a considerable change being introduced.
This changes has been upvoted by community and @ajcvickers mentioned in earlier PRs that the repo consolidations were part of the blocker for this fix to getting merged in.

Also adding @HaoK to this PR. Do you have any major hesitations on this enhancement getting introduced into extensions configuration or it sounds good to you?

amoerie commented:
> We'll need to do a full review on the changes as it seems like there is a considerable change being introduced.

I just want to add a little note to this sentence: the internal changes are indeed considerable, but the external compatibility has not been broken. I did not modify any existing tests (except for 2 tests that verified duplicate keys are not supported - now they are) and only added tests to verify the new supported behavior. 

In any case, I'm very interested to hear your feedback on my code.

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs:
```diff
@@ -42,57 +42,61 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
-                            prefixStack.Push(reader.LocalName);
-                            ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                            ProcessAttributes(reader, prefixStack, data, AddAttributePair);
+                            var parent = currentPath.Any() ? currentPath.Peek() : null;
```
nit this line and some other instances: `We only use var when it's obvious what the variable type is (e.g. var stream = new FileStream(...) not var stream = OpenStandardInput()).`

taken from [coding-style.md](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md)

amoerie replied:
Ah I actually perused through the code style but I must have missed this. 
Updated!

maryamariyan commented on src/libraries/Microsoft.Extensions.Configuration.Xml/tests/ConfigurationProviderXmlTest.cs:
```diff
@@ -20,6 +20,16 @@ public override void Combine_after_other_provider()
             // Disabled test due to XML handling of empty section.
         }
 
+        public override void Load_from_single_provider_with_duplicates_throws()
```
Should we be doing the same thing as below instead?
https://github.com/dotnet/runtime/blob/8fcb543625324c81f934e67a15077ad903dd3c9f/src/libraries/Microsoft.Extensions.Configuration.Json/tests/ConfigurationProviderJsonTest.cs#L13-L19

and 

https://github.com/dotnet/runtime/blob/8fcb543625324c81f934e67a15077ad903dd3c9f/src/libraries/Microsoft.Extensions.Configuration.CommandLine/tests/ConfigurationProviderCommandLineTest.cs#L38-L49

amoerie replied:
Well the behavior with duplicate keys is already covered by other tests. These tests in the JSON provider simply test if the provider does not crash on duplicate keys, while my added tests also verify if the actual output is correct.

But when in doubt, it's always better to be consistent. I've followed the JSON example and implemented these tests the same for the XML provider.

amoerie committed:
Fix code style violations: only use  var when the type is obvious

amoerie committed:
Run duplicate test config through the XML provider

This ensures that, at the very least, it doesn't crash
Correct behavior is covered by other tests

HaoK commented:
I don't have any particular concerns, other than the general one around accidental breaking changes.  This behavior in particular has been asked for frequently, and it was always near the top of the list of PRs we were considering, so I think its a good candidate to be worth the risk.

HaoK commented:
One thing worth mentioning is that this change appears only to be made in the XmlStreamProvider, it would be strange for there to be a big behavior difference between the stream and file providers?

amoerie commented:
> One thing worth mentioning is that this change appears only to be made in the XmlStreamProvider, it would be strange for there to be a big behavior difference between the stream and file providers?

Are you saying there's other config provider implementations for XML files somewhere? Could you point me to them? I was operating under the assumption that everything would use the classes that I modified.

amoerie commented:
Small update: I went digging a little and found no other implementations for XML files. In fact, I am convinced now that XML files also simply pass through the XmlStreamConfigurationProvider. See this class:

https://github.com/dotnet/runtime/blob/f3d84ca003943b09a8b80091e4b61221be339b22/src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationProvider.cs#L8-L29

amoerie commented:
/azp run runtime

maryamariyan commented:
/azp run runtime

amoerie commented:
> /azp run runtime

Thanks @maryamariyan 

Perhaps the [PR guide](https://github.com/dotnet/runtime/blob/master/docs/pr-guide.md) could use an update, because I didn't see any option aside from pushing more commits or closing and reopening this PR.

amoerie commented:
I'll close and reopen this PR to avoid merge commits.

maryamariyan commented:
@amoerie how about the "rerun all checks" tab?
I think it refers to the "rerun" links in https://github.com/dotnet/runtime/pull/43722/checks?check_run_id=1302103237

safern commented:
I believe that is because that is not available for contributors, only for collaborators. I believe write permissions are needed in order to use /azp run... and to see the rerun buttons on the checks tab.

maryamariyan commented:
> One thing worth mentioning is that this change appears only to be made in the XmlStreamProvider, it would be strange for there to be a big behavior difference between the stream and file providers?

> Small update: I went digging a little and found no other implementations for XML files. In fact, I am convinced now that XML files also simply pass through the XmlStreamConfigurationProvider.

We have three file providers, and XmlConfigurationProvider is one of them, the other two are JsonConfigurationProvider and IniConfigurationProvider.

If I understood correctly @HaoK recommended if we are going to accept this breaking change in behavior that should also be done for the other two.

HaoK commented:
Actually I wasn't talking about the other types of provider, I was referring to the old XmlConfigurationProvider, but indeed its just routing calls through the same code path so we should be good.

https://github.com/amoerie/runtime/blob/88e4f545e95d6952fbb34926f1dc9156b4e0baf6/src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlConfigurationProvider.cs#L27

amoerie commented:
I've recreated my pull request here: #44608

This cleans up the commit history (i've squashed my commits into one) and properly retries the CI.
